README.md
<img src="https://i.imgur.com/k0bVKkM.png" width="550">

# Trello Watchman
![Python 2.7 and 3 compatible](https://img.shields.io/pypi/pyversions/trello-watchman)
![PyPI version](https://img.shields.io/pypi/v/trello-watchman.svg)
![License: MIT](https://img.shields.io/pypi/l/trello-watchman.svg)

Monitoring your Trello boards for sensitive information

## About Trello Watchman
Trello Watchman is an application that uses the Trello API to look for potentially sensitive data exposed in your public Trello boards.

### Features
Trello Watchman looks for:

- API Keys, Tokens & Service Accounts
  - AWS, Azure, GCP, Google API, Slack (keys & webhooks), Twitter, Facebook, GitHub
  - Generic Private keys
  - Access Tokens, Bearer Tokens, Client Secrets, Private Tokens
- Files
    - Certificate files
    - Potentially interesting/malicious/sensitive files (.docm, .xlsm, .zip etc.)
    - Executable files
- Personal Data
    - Leaked passwords
    - Passport numbers, Dates of birth, Social security numbers, National insurance numbers, Drivers licence numbers (UK), Individual Taxpayer Identification Number

#### Time based searching
You can run Trello Watchman to look for results going back as far as:
- 24 hours
- 7 days
- 30 days
- All time

This means after one deep scan, you can schedule Trello Watchman to run regularly and only return results from your chosen timeframe.

### Rules
Trello Watchman uses custom YAML rules to detect matches in Trello.

They follow this format:

```yaml
---
filename:
enabled: [true|false]
meta:
  name:
  author:
  date:
  description: #what the search should find
  severity: #rating out of 100
scope:
- #[text|attachments]
test_cases:
  match_cases:
  - #test case that should match the regex*
  fail_cases:
  - #test case that should not match the regex*
strings:
- #search query to use in Trello*
pattern: #Regex pattern to filter out false positives*
```
There are Python tests to ensure rules are formatted properly and that the Regex patterns work in the `tests` dir

More information about rules, and how you can add your own, is in the file `docs/rules.md`.

### Logging

Trello Watchman gives the following logging options:
- Stdout
- Log file
- TCP stream

Results are output in JSON format, perfect for ingesting into a SIEM or other log analysis platform.

For file and TCP stream logging, configuration options need to be passed via `.conf` file or environment variable. See the file `docs/logging.md` for instructions on how to set it up.

If no logging option is given, Trello Watchman defaults to Stdout logging.

## Requirements
### Trello API token
To run Trello Watchman, you will need a Trello API OAuth access token, which take the form of a `key` and a `secret`. You can generate these [here](https://trello.com/app-key).

**Note**: User tokens act on behalf of the user who generates them, so I would suggest you create this app and authorise it using a service account, otherwise the app will have access to your private Trello boards.

#### Providing token
Trello Watchman will first try to get the secret and key from the environment variables `TRELLO_WATCHMAN_SECRET` and `TRELLO_WATCHMAN_KEY`, if this fails it will load the token from .conf file (see below).

### .conf file
Configuration options can be passed in a file named `watchman.conf` which must be stored in your home directory. The file should follow the YAML format, and should look like below:
```yaml
trello_watchman:
  secret: abc123
  key: abc123
  logging:
    file_logging:
      path: /var/log/
    json_tcp:
      host: localhost
      port: 9020
```
Trello Watchman will look for this file at runtime, and use the configuration options from here. If you are not using the advanced logging features, leave them blank.

If you are having issues with your .conf file, run it through a YAML linter.

An example file is in `docs/example.conf`

## Installation
Install via pip

`python3 -m pip install trello-watchman`

## Usage
Trello Watchman will be installed as a global command, use as follows:
```
usage: trello-watchman [-h] --timeframe {d,w,m,a} [--output {file,stdout,stream}]
                   [--version] [--all] [--attachments] [--text]

Monitoring your Trello boards for sensitive information

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit
  --all                 Find everything
  --attachments         Search for attachments
  --text                Search text

required arguments:
  --timeframe {d,w,m,a}
                        How far back to search: d = 24 hours w = 7 days, m =
                        30 days, a = all time
  --output {file,stdout,stream}
                        Where to send results

  ```

You can run Trello Watchman to look for everything, and output to default CSV:

`trello-watchman --timeframe a --all`

## Other Watchman apps
You may be interested in some of the other apps in the Watchman family:
- [Slack Watchman](https://github.com/PaperMtn/slack-watchman)
- [GitLab Watchman](https://github.com/PaperMtn/gitlab-watchman)
- [GitHub Watchman](https://github.com/PaperMtn/github-watchman)

## License
The source code for this project is released under the [GNU General Public Licence](https://www.gnu.org/licenses/licenses.html#GPL). This project is not associated with Trello or Atlassian.


CHANGELOG.md
## 1.0.1 - 2021-05-07
Release

### Added
- Actions for testing and publishing added
- Supported Python versions updated



tests/test_rules.py
import yaml
import os
import unittest
from pathlib import Path

from trello_watchman import rule

RULES_PATH = (Path(__file__).parent / 'rules').resolve()


def load_rules() -> list:
    """Load rules from YAML files

    Returns:
        List containing loaded definitions as Rule objects
    """

    loaded_definitions = []
    try:
        for root, dirs, files in os.walk(RULES_PATH):
            for rule_file in files:
                rule_path = (Path(root) / rule_file).resolve()
                if rule_path.name.endswith('.yaml'):
                    loaded_definitions.append(rule.load_from_yaml(rule_path))
        return loaded_definitions
    except Exception as e:
        raise e


def check_yaml(rule):
    try:
        yaml_rule = yaml.safe_load(rule)
    except:
        return False
    return True


class TestEmail(unittest.TestCase):
    def test_rule_format(self):
        """Check rules are properly formed YAML ready to be ingested"""

        for root, dirs, files in os.walk(RULES_PATH):
            for rule_file in files:
                rule_path = (Path(root) / rule_file).resolve()
                if rule_path.name.endswith('.yaml'):
                    with open(rule_path) as yaml_file:
                        self.assertTrue(check_yaml(yaml_file.read()), msg=f'Malformed YAML: {yaml_file.name}')

    def test_rule_matching_cases(self):
        """Test that the match case strings match the regex. Skip if the match case is 'blank'"""

        rules_list = load_rules()
        for rule in rules_list:
            for test_case in rule.test_cases.match_cases:
                if not test_case == 'blank':
                    self.assertRegex(test_case, rule.pattern, msg='Regex does not detect given match case')

    def test_rule_failing_cases(self):
        """Test that the fail case strings don't match the regex. Skip if the fail case is 'blank'"""

        rules_list = load_rules()
        for rule in rules_list:
            for test_case in rule.test_cases.fail_cases:
                if not test_case == 'blank':
                    self.assertNotRegex(test_case, rule.pattern,
                                        msg='Regex does detect given failure case, it should '
                                            'not')


if __name__ == '__main__':
    unittest.main()


trello_watchman/rule.py
import pathlib
import yaml
from collections import namedtuple


class Rule(object):
    """Class that handles loaded rule objects"""

    def __init__(self,
                 filename: str,
                 enabled: bool,
                 meta: namedtuple,
                 scope: list,
                 test_cases: namedtuple,
                 strings: str,
                 pattern: str):
        self.filename = filename
        self.enabled = enabled
        self.meta = meta
        self.scope = scope
        self.test_cases = test_cases
        self.strings = strings
        self.pattern = pattern

    def __repr__(self):
        return f'{self.__class__.__name__}({self.__dict__!r})'

    def __str__(self):
        return ' '.join(f'{k}: {v!s}' for k, v in self.__dict__.items())


def load_from_yaml(rule_path: pathlib.PosixPath) -> Rule:
    """Load YAML file and return a Rule object

    Args:
        rule_path: Path of YAML file
    Returns:
        Definition object with fields populated from the YAML
        definition file
    """

    with open(rule_path) as yaml_file:
        yaml_import = yaml.safe_load(yaml_file)

        meta = namedtuple('meta', ['name', 'author', 'date', 'version', 'description', 'severity'])
        meta.name = yaml_import.get('meta').get('name')
        meta.author = yaml_import.get('meta').get('author')
        meta.date = yaml_import.get('meta').get('date')
        meta.version = yaml_import.get('meta').get('version')
        meta.description = yaml_import.get('meta').get('description')
        meta.severity = yaml_import.get('meta').get('severity')

        test_cases = namedtuple('test_cases', ['match_cases', 'fail_cases'])
        test_cases.match_cases = yaml_import.get('test_cases').get('match_cases')
        test_cases.match_cases = yaml_import.get('test_cases').get('fail_cases')

        rule = Rule(filename=yaml_import.get('filename'),
                    enabled=yaml_import.get('enabled'),
                    meta=meta,
                    scope=yaml_import.get('scope'),
                    test_cases=test_cases,
                    strings=yaml_import.get('strings'),
                    pattern=yaml_import.get('pattern'))
    return rule


trello_watchman/logger.py
import os
import logging
import socket
import sys
import logging.handlers
import simplejson as json
from datetime import datetime
from logging import Logger


class LoggingBase(Logger):
    def __init__(self, name='Trello Watchman'):
        super().__init__(name)
        self.notify_format = logging.Formatter(
            '{"localtime": "%(asctime)s", "level": "NOTIFY", "source": "%(name)s", "scope": "%(scope)s",'
            ' "severity": "%(severity)s", "detection_type": "%(type)s", "detection_data": %(message)s}')
        self.info_format = logging.Formatter(
            '{"localtime": "%(asctime)s", "level": "%(levelname)s", "source": "%(name)s", "message":'
            ' "%(message)s"}')
        self.log_path = ''
        self.logger = logging.getLogger(self.name)
        self.logger.setLevel(logging.DEBUG)


class FileLogger(LoggingBase):
    def __init__(self, log_path):
        LoggingBase.__init__(self)
        self.handler = logging.handlers.WatchedFileHandler(os.path.join(log_path, 'trello_watchman.log'))
        self.logger.addHandler(self.handler)

    def log_notification(self, log_data, scope, detect_type, severity):
        self.handler.setFormatter(self.notify_format)
        self.logger.warning(json.dumps(log_data), extra={
            'scope': scope,
            'type': detect_type,
            'severity': severity
        })

    def log_info(self, log_data):
        self.handler.setFormatter(self.info_format)
        self.logger.info(log_data)

    def log_critical(self, log_data):
        self.handler.setFormatter(self.info_format)
        self.logger.critical(log_data)


class StdoutLogger(LoggingBase):
    def __init__(self):
        LoggingBase.__init__(self)
        self.handler = logging.StreamHandler(sys.stdout)
        self.logger.addHandler(self.handler)

    def log_notification(self, log_data, scope, detect_type, severity):
        self.handler.setFormatter(self.notify_format)
        self.logger.warning(json.dumps(log_data), extra={
            'scope': scope,
            'type': detect_type,
            'severity': severity
        })

    def log_info(self, log_data):
        self.handler.setFormatter(self.info_format)
        self.logger.info(log_data)

    def log_critical(self, log_data):
        self.handler.setFormatter(self.info_format)
        self.logger.critical(log_data)


class SocketJSONLogger(object):
    def __init__(self, host, port):
        self.host = host
        self.port = port
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            self.sock.connect((self.host, self.port))
        except socket.error as error:
            print(error)

    def send(self, data):
        try:
            self.sock.sendall(bytes(data, encoding="utf-8"))
        except Exception as e:
            print(e)

    def log_notification(self, log_data, scope, detect_type, severity):
        message = json.dumps({
            'localtime': datetime.now().strftime('%Y-%m-%d %H:%M:%S,%f'),
            'level': 'NOTIFY',
            'source': 'Trello Watchman',
            'scope': scope,
            'severity': severity,
            'detection_type': detect_type,
            'detection_data': log_data
        }) + '\n'
        self.send(message)

    def log_info(self, log_data):
        message = json.dumps({
            'localtime': datetime.now().strftime('%Y-%m-%d %H:%M:%S,%f'),
            'level': 'INFO',
            'source': 'Trello Watchman',
            'message': log_data
        }) + '\n'
        self.send(message)

    def log_critical(self, log_data):
        message = json.dumps({
            'localtime': datetime.now().strftime('%Y-%m-%d %H:%M:%S,%f'),
            'level': 'CRITICAL',
            'source': 'Trello Watchman',
            'message': log_data
        }) + '\n'
        self.send(message)

trello_watchman/__about__.py
# Trello Watchman
# Copyright (C) 2021  PaperMtn
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

__all__ = [
    '__title__',
    '__summary__',
    '__uri__',
    '__version__',
    '__author__',
    '__email__',
    '__license__',
]

__title__ = 'Trello Watchman'
__summary__ = 'Monitoring your Trello boards for sensitive information'
__uri__ = 'https://github.com/PaperMtn/trello-watchman'
__version__ = '0.0.9'
__author__ = 'PaperMtn'
__email__ = 'papermtn@protonmail.com'
__license__ = 'GPL-3.0'
__copyright__ = f'2021 {__author__}'


trello_watchman/__init__.py
import calendar
import time
import argparse
import builtins
import os
import yaml
from pathlib import Path

from trello_watchman import __about__
from trello_watchman import trello_wrapper
from trello_watchman import logger
from trello_watchman import rule

DAY_TIMEFRAME = 86400
MONTH_TIMEFRAME = 2592000
WEEK_TIMEFRAME = 604800
ALL_TIME = calendar.timegm(time.gmtime()) + 1576800000

RULES_PATH = (Path(__file__).parent / 'rules').resolve()
OUTPUT_LOGGER = ''


def load_rules() -> list:
    """Load rules from YAML files

    Returns:
        List containing loaded definitions as Rule objects
    """

    loaded_definitions = []
    try:
        for root, dirs, files in os.walk(RULES_PATH):
            for rule_file in files:
                rule_path = (Path(root) / rule_file).resolve()
                if rule_path.name.endswith('.yaml'):
                    loaded_definitions.append(rule.load_from_yaml(rule_path))
        return loaded_definitions
    except Exception as e:
        raise e


def validate_conf(path: str) -> bool or list:
    """Check the file watchman.conf exists

    Args:
        path: Path where the conf file should be
    Returns:
        Either bool if environment variables have been set
        or True if the file watchman.conf file exists in the given path
    """

    if os.environ.get('TRELLO_WATCHMAN_SECRET') and os.environ.get('TRELLO_WATCHMAN_KEY'):
        return True
    if os.path.exists(path):
        with open(path) as yaml_file:
            return yaml.safe_load(yaml_file).get('trello_watchman')


def search(trello_conn: trello_wrapper.TrelloAPI, rule: rule.Rule, tf: int, scope: str):
    """Carries out a search on the Trello API based on the given rule,
     timeframe and scope

        Args:
            trello_conn: Trello API connection object to carry out the search
            rule: Rule object to use for the search
            tf: Epoch timeframe to search back in
            scope: What to search Trello for
        """

    if isinstance(OUTPUT_LOGGER, logger.StdoutLogger):
        print = OUTPUT_LOGGER.log_info
    else:
        print = builtins.print

    if scope == 'attachments':
        print(f'Searching for attachments containing {rule.meta.name}')
        attachments = trello_wrapper.find_attachments(trello_conn, OUTPUT_LOGGER, rule, tf)
        if attachments:
            for log_data in attachments:
                OUTPUT_LOGGER.log_notification(log_data, scope, rule.meta.name,
                                               rule.meta.severity)
    if scope == 'text':
        print(f'Searching for cards containing {rule.meta.name}')
        text = trello_wrapper.find_text(trello_conn, OUTPUT_LOGGER, rule, tf)
        if text:
            for log_data in text:
                OUTPUT_LOGGER.log_notification(log_data, scope, rule.meta.name,
                                               rule.meta.severity)


def main():
    global OUTPUT_LOGGER

    if isinstance(OUTPUT_LOGGER, logger.StdoutLogger):
        print = OUTPUT_LOGGER.log_critical
    else:
        print = builtins.print

    try:
        parser = argparse.ArgumentParser(description=__about__.__summary__)
        required = parser.add_argument_group('required arguments')
        required.add_argument('--timeframe', choices=['d', 'w', 'm', 'a'], dest='time',
                              help='How far back to search: d = 24 hours w = 7 days, m = 30 days, a = all time',
                              required=True)
        required.add_argument('--output', choices=['file', 'stdout', 'stream'], dest='logging_type',
                              help='Where to send results')
        parser.add_argument('--version', action='version',
                            version=f'trello-watchman {__about__.__version__}')
        parser.add_argument('--all', dest='everything', action='store_true',
                            help='Find everything')
        parser.add_argument('--attachments', dest='attachments', action='store_true',
                            help='Search for attachments')
        parser.add_argument('--text', dest='text', action='store_true',
                            help='Search text')

        args = parser.parse_args()
        tm = args.time
        everything = args.everything
        attachments = args.attachments
        text = args.text
        logging_type = args.logging_type

        if tm == 'd':
            tf = DAY_TIMEFRAME
        elif tm == 'w':
            tf = WEEK_TIMEFRAME
        elif tm == 'm':
            tf = MONTH_TIMEFRAME
        else:
            tf = ALL_TIME

        conf_path = f'{os.path.expanduser("~")}/watchman.conf'

        if not validate_conf(conf_path):
            raise Exception(f'TRELLO_WATCHMAN_SECRET/TRELLO_WATCHMAN_KEY environment variable or watchman.conf file '
                            f'not detected. \nEnsure environment variable is set or a valid file is located in your '
                            f'home directory: {os.path.expanduser("~")}')
        else:
            config = validate_conf(conf_path)
            connection = trello_wrapper.initiate_trello_connection()

        if logging_type:
            if logging_type == 'file':
                if os.environ.get('TRELLO_WATCHMAN_LOG_PATH'):
                    OUTPUT_LOGGER = logger.FileLogger(os.environ.get('TRELLO_WATCHMAN_LOG_PATH'))
                elif config.get('logging').get('file_logging').get('path') and \
                        os.path.exists(config.get('logging').get('file_logging').get('path')):
                    OUTPUT_LOGGER = logger.FileLogger(log_path=config.get('logging').get('file_logging').get('path'))
                else:
                    print('No config given, outputting trello_watchman.log file to home path')
                    OUTPUT_LOGGER = logger.FileLogger(log_path=os.path.expanduser('~'))
            elif logging_type == 'stdout':
                OUTPUT_LOGGER = logger.StdoutLogger()
            elif logging_type == 'stream':
                if os.environ.get('TRELLO_WATCHMAN_HOST') and os.environ.get('TRELLO_WATCHMAN_PORT'):
                    OUTPUT_LOGGER = logger.SocketJSONLogger(os.environ.get('TRELLO_WATCHMAN_HOST'),
                                                            os.environ.get('TRELLO_WATCHMAN_PORT'))
                elif config.get('logging').get('json_tcp').get('host') and \
                        config.get('logging').get('json_tcp').get('port'):
                    OUTPUT_LOGGER = logger.SocketJSONLogger(config.get('logging').get('json_tcp').get('host'),
                                                            config.get('logging').get('json_tcp').get('port'))
                else:
                    raise Exception("JSON TCP stream selected with no config")
        else:
            print('No logging option selected, defaulting to stdout')
            OUTPUT_LOGGER = logger.StdoutLogger()

        if not isinstance(OUTPUT_LOGGER, logger.StdoutLogger):
            print = builtins.print
            print('Trello Watchman')
            print(f'Version: {__about__.__version__}\n')
            print('Importing rules...')
            rules_list = load_rules()
            print(f'{len(rules_list)} rules loaded')
        else:
            OUTPUT_LOGGER.log_info(f'Trello Watchman started execution - Version: {__about__.__version__}')
            OUTPUT_LOGGER.log_info('Importing rules...')
            rules_list = load_rules()
            OUTPUT_LOGGER.log_info(f'{len(rules_list)} rules loaded')
            print = OUTPUT_LOGGER.log_info

        if everything:
            print('Getting everything...')
            for rule in rules_list:
                if 'attachments' in rule.scope:
                    search(connection, rule, tf, 'attachments')
                if 'text' in rule.scope:
                    search(connection, rule, tf, 'text')
        else:
            if attachments:
                print('Getting attachments')
                for rule in rules_list:
                    if 'attachments' in rule.scope:
                        search(connection, rule, tf, 'attachments')
            if text:
                print('Getting cards')
                for rule in rules_list:
                    if 'text' in rule.scope:
                        search(connection, rule, tf, 'text')

        print('++++++Audit completed++++++')

    except Exception as e:
        if isinstance(OUTPUT_LOGGER, logger.StdoutLogger):
            print = OUTPUT_LOGGER.log_critical
        else:
            print = builtins.print

        print(e)


if __name__ == '__main__':
    main()
